File: in-app-messaging-java/app/src/main/java/com/vonage/tutorial/messaging/ChatFragment.java
package com.vonage.tutorial.messaging;

import android.os.Bundle;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.*;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.fragment.NavHostFragment;
import com.nexmo.client.NexmoEvent;
import com.nexmo.client.NexmoMemberEvent;
import com.nexmo.client.NexmoTextEvent;

import java.util.ArrayList;

public class ChatFragment extends Fragment implements BackPressHandler {

    private ChatViewModel viewModel;

    private ProgressBar progressBar;
    private TextView errorTextView;
    private ConstraintLayout chatContainer;
    private Button logoutButton;
    private Button sendMessageButton;
    private TextView userNameTextView;
    private EditText messageEditText;
    private TextView conversationEventsTextView;

    private Observer<String> errorMessageObserver = it -> {
        progressBar.setVisibility(View.GONE);
        errorTextView.setText(it);

        if (it.equals("")) {
            errorTextView.setVisibility(View.GONE);
            chatContainer.setVisibility(View.VISIBLE);
        } else {
            errorTextView.setVisibility(View.VISIBLE);
            chatContainer.setVisibility(View.GONE);
        }
    };

    private Observer<String> userNameObserver = it -> {
        userNameTextView.setText(it + ": ");
        logoutButton.setText("Logout " + it);
    };

    private Observer<ArrayList<NexmoEvent>> conversationEvents = events -> {

        ArrayList<String> lines = new ArrayList<>();

        for (NexmoEvent event : events) {
            if (event == null) {
                continue;
            }

            String line = "";

            if (event instanceof NexmoMemberEvent) {
                line = getConversationLine((NexmoMemberEvent) event);
            } else if (event instanceof NexmoTextEvent) {
                line = getConversationLine((NexmoTextEvent) event);
            }

            lines.add(line);
        }

        // Production application should utilise RecyclerView to provide better UX
        if (events.isEmpty()) {
            conversationEventsTextView.setText("Conversation has no events");
        } else {

            String conversationEvents = "";

            for (String line : lines) {
                conversationEvents += line + "\n";
            }

            conversationEventsTextView.setText(conversationEvents);
        }

        progressBar.setVisibility(View.GONE);
        chatContainer.setVisibility(View.VISIBLE

        );
    };

    public ChatFragment() {
        super(R.layout.fragment_chat);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        progressBar = view.findViewById(R.id.progressBar);
        errorTextView = view.findViewById(R.id.errorTextView);
        chatContainer = view.findViewById(R.id.chatContainer);
        logoutButton = view.findViewById(R.id.logoutButton);
        sendMessageButton = view.findViewById(R.id.sendMessageButton);
        userNameTextView = view.findViewById(R.id.userNameTextView);
        messageEditText = view.findViewById(R.id.messageEditText);
        conversationEventsTextView = view.findViewById(R.id.conversationEventsTextView);

        viewModel = new ViewModelProvider(requireActivity()).get(ChatViewModel.class);

        if (Config.CONVERSATION_ID.trim().isEmpty()) {
            Toast.makeText(requireActivity(), "Please set Config.CONVERSATION_ID", Toast.LENGTH_SHORT).show();
            onBackPressed();
            return;
        }

        viewModel.onInit();

        viewModel.errorMessage.observe(getViewLifecycleOwner(), errorMessageObserver);
        viewModel.conversationEvents.observe(getViewLifecycleOwner(), conversationEvents);
        viewModel.userName.observe(getViewLifecycleOwner(), userNameObserver);

        sendMessageButton.setOnClickListener(it -> {
            String message = messageEditText.getText().toString();

            if (!message.trim().isEmpty()) {
                viewModel.onSendMessage(messageEditText.getText().toString());
                messageEditText.setText("");
                hideKeyboard();
            } else {
                Toast.makeText(requireActivity(), "Message is blank", Toast.LENGTH_SHORT).show();
            }
        });

        logoutButton.setOnClickListener(it -> {
            viewModel.onLogout();
            NavHostFragment.findNavController(this).popBackStack();
        });
    }

    private String getConversationLine(NexmoMemberEvent memberEvent) {
        String user = memberEvent.getEmbeddedInfo().getUser().getName();

        switch (memberEvent.getState()) {
            case JOINED:
                return user + " joined";
            case INVITED:
                return user + " invited";
            case LEFT:
                return user + " left";
            case UNKNOWN:
                return "Error: Unknown member event state";
        }

        return "";
    }

    private String getConversationLine(NexmoTextEvent textEvent) {
        String user = textEvent.getEmbeddedInfo().getUser().getName();
        return user + "  said: " + textEvent.getText();
    }

    public void hideKeyboard() {
        InputMethodManager inputMethodManager = ContextCompat.getSystemService(getContext(), InputMethodManager.class);

        View view = getActivity().getCurrentFocus();

        if (view == null) {
            view = new View(getActivity());
        }

        inputMethodManager.hideSoftInputFromWindow(view.getWindowToken(), 0);
    }

    @Override
    public void onBackPressed() {
        viewModel.onBackPressed();
    }
}

File: in-app-messaging-kotlin/app/src/main/java/com/vonage/tutorial/messaging/ChatFragment.kt
package com.vonage.tutorial.messaging

import android.os.Bundle
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.Button
import android.widget.EditText
import android.widget.ProgressBar
import android.widget.TextView
import android.widget.Toast
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.core.content.ContextCompat
import androidx.core.view.isVisible
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.Observer
import androidx.navigation.fragment.findNavController
import com.nexmo.client.NexmoEvent
import com.nexmo.client.NexmoMemberEvent
import com.nexmo.client.NexmoMemberState
import com.nexmo.client.NexmoTextEvent

class ChatFragment : Fragment(R.layout.fragment_chat), BackPressHandler {

    private val viewModel by viewModels<ChatViewModel>()

    private lateinit var progressBar: ProgressBar
    private lateinit var errorTextView: TextView
    private lateinit var chatContainer: ConstraintLayout
    private lateinit var logoutButton: Button
    private lateinit var sendMessageButton: Button
    private lateinit var userNameTextView: TextView
    private lateinit var messageEditText: EditText
    private lateinit var conversationEventsTextView: TextView

    private var errorMessageObserver = Observer<String> {
        progressBar.isVisible = false
        errorTextView.text = it
        errorTextView.isVisible = it.isNotEmpty()
        chatContainer.isVisible = it.isEmpty()
    }

    private var userNameObserver = Observer<String> {
        userNameTextView.text = "$it says:"
        logoutButton.text = "Logout $it"
    }

    private var conversationEvents = Observer<List<NexmoEvent>?> { events ->
        val messages = events?.mapNotNull {
            when (it) {
                is NexmoMemberEvent -> getConversationLine(it)
                is NexmoTextEvent -> getConversationLine(it)
                else -> null
            }
        }

        conversationEventsTextView.text = if (messages.isNullOrEmpty()) {
            "Conversation has No messages"
        } else {
            messages.joinToString(separator = "\n")
        }

        progressBar.isVisible = false
        chatContainer.isVisible = true
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        if (Config.CONVERSATION_ID.isBlank()) {

            Toast.makeText(context, "Please set Config.CONVERSATION_ID", Toast.LENGTH_SHORT).show()
            activity?.onBackPressed()
            return
        }

        viewModel.onInit()

        viewModel.errorMessage.observe(viewLifecycleOwner, errorMessageObserver)
        viewModel.conversationEvents.observe(viewLifecycleOwner, conversationEvents)
        viewModel.userName.observe(viewLifecycleOwner, userNameObserver)

        progressBar = view.findViewById(R.id.progressBar)
        errorTextView = view.findViewById(R.id.errorTextView)
        chatContainer = view.findViewById(R.id.chatContainer)
        logoutButton = view.findViewById(R.id.logoutButton)
        sendMessageButton = view.findViewById(R.id.sendMessageButton)
        userNameTextView = view.findViewById(R.id.userNameTextView)
        messageEditText = view.findViewById(R.id.messageEditText)
        conversationEventsTextView = view.findViewById(R.id.conversationEventsTextView)

        sendMessageButton.setOnClickListener {
            val message = messageEditText.text.toString()

            if (message.isNotBlank()) {
                viewModel.onSendMessage(messageEditText.text.toString())
                messageEditText.setText("")
                hideKeyboard()
            } else {
                Toast.makeText(context, "Message is blank", Toast.LENGTH_SHORT).show()
            }
        }

        logoutButton.setOnClickListener {
            viewModel.onLogout()
            findNavController().popBackStack()
        }
    }

    private fun getConversationLine(memberEvent: NexmoMemberEvent): String {
        val user = memberEvent.embeddedInfo.user.name

        return when (memberEvent.state) {
            NexmoMemberState.JOINED -> "$user joined"
            NexmoMemberState.INVITED -> "$user invited"
            NexmoMemberState.LEFT -> "$user left"
            else -> "Error: Unknown member event state"
        }
    }

    private fun getConversationLine(textEvent: NexmoTextEvent): String {
        val user = textEvent.embeddedInfo.user.name
        return "$user said: ${textEvent.text}"
    }

    override fun onBackPressed() {
        viewModel.onBackPressed()
    }

    private fun hideKeyboard() {
        val context = context ?: return

        val inputMethodManager = ContextCompat.getSystemService(context, InputMethodManager::class.java)
        var view = activity?.currentFocus
        if (view == null) {
            view = View(activity)
        }

        inputMethodManager?.hideSoftInputFromWindow(view.windowToken, 0)
    }
}
